### PR TITLE
Completed lesson03 for John Rogers.

### PR DESCRIPTION
**Pull request by johnroge**
Completed mailroom.py and strformat_lab.py for lesson03, john_rogers.

### CODE REVIEW

File: students/john_rogers/lesson03/mailroom.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    # Not the prettiest formatting, but works well enough for v1
+    for donor in data:
+        total_amount = round(float(sum(data[donor].values())), 2)
+        num_donations = round(len(data[donor].values()), 2)
+        avg_donation = total_amount / num_donations
+        avg_donation = round(avg_donation, 2)
+        dollar = '$'
+
+        print(f'{donor.capitalize():20} {dollar:1} {total_amount:3n}'
+              f' {num_donations:12n} {avg_donation:12n}')
```
[the remark this answers is not in the document]
Comment (johnroge) [reply students/john_rogers/lesson03/mailroom.py]: Resolved.

File: students/john_rogers/lesson03/mailroom.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                    'd4': 456.23,
+                    },
+         }
+
+    while True:
+        choice = get_input()
+
+        if choice == 1:
+            print('Exiting program.')
+            exit()
```
[the remark this answers is not in the document]
Comment (johnroge) [reply students/john_rogers/lesson03/mailroom.py]: Changed to 'break' and tested, resolved.

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    # make some global variables to experiment with
+    global db
+    global main_prompt
+    global main_dispatch
+
+    # create a main database
+    db = {'sting': {'d1': 13.45,
+                    'd2': 214.34,
+                    'd3': 453.23,
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: what are d1, d2, and d3?

if you intend to use nested dict then use clear dict keys like "total", "donations" etc or simplify to just have donations as value:
`donors = {"bob": [1, 3, 4]}`

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        ">>>\n"
+    )
+
+    # create the main menu using dict switch
+    main_dispatch = {
+        '1': exit_menu,
+        '2': thank_you,
+        '3': create_report,
+        '4': write_report,
+    }
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: 👍

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+def menu():
+    """
+    Get user input
+    :return: call the appropriate menu item
+    """
+    while True:
+        response = input(main_prompt)
+        if response == '1':
+            exit_menu()
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: why special if statement? you have "1" mapped to this function in `main_dispatch`

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    pass
+
+
+def exit_menu():
+    """
+    exit the program
+    :return: SystemExit
+    """
+    print('Exiting program -')
+    raise SystemExit
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: use `sys.exit()` (which is really the same underneath)

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    global db
+    global main_prompt
+    global main_dispatch
+
+    # create a main database
+    db = {'sting': {'d1': 13.45,
+                    'd2': 214.34,
+                    'd3': 453.23,
+                    },
+          'bono': {'d1': 54.54,
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: reading your code more.. are d1, d2 just positions for the list of donations? if so, what benefit does it give you vs just a list of donations?

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        else:
+            # add the new name and prompt for donation
+            print(f'Welcome aboard, {cmd.capitalize()}, how much would '
+                  f'you like to donate?')
+            new_donation = input('Please an amount to donate: ')
+            new_donation = float(new_donation)
+            names[cmd] = {'d1': new_donation}
+
+            print(f'Thank you, {cmd.capitalize()}, for your kind'
+                  f' donation of ${new_donation}.')
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: there is quite a bit of duplicate code in the lines 128 to 149 - I would recommend refactoring that.

File: students/john_rogers/lesson04/mailroom_pt2.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+def create_report(data):
+    """
+    Print a list of donors sorted by historical donation amount.
+    List donor name, number of donations and average donation amount.
+    :return: none
+    """
+
+    # Create a list from the donor dictionary so we can sort by size
+    d_list = []
+    for donor in data:
```
Comment (natasha-aleksandrova) [inline students/john_rogers/lesson04/mailroom_pt2.py]: it is better practice to use items -> `for name, donations in data.items()`

**Pull request merged**